src/syntax/FStar.Syntax.Syntax.fsti

```diff
@@ -429,39 +449,41 @@ type wp_eff_combinators = {
 (*
  * Layered effects combinators
  *
- * All of these are pairs of type schemes,
+ * All of these have pairs of type schemes,
  *   where the first component is the term ts and the second component is the type ts
  *
  * Before typechecking the effect declaration, the second component is a dummy ts
  *   In other words, desugaring sets the first component only, and typechecker then fills up the second one
  *
- * Similarly the base effect name is also "" after desugaring, and is set by the typechecker
+ * Additionally, bind, subcomp, and if_then_else have a combinator kind,
+ *   this is also set to None in desugaring and set during typechecking the effect
  *)
 type layered_eff_combinators = {
   l_repr         : (tscheme * tscheme);
   l_return       : (tscheme * tscheme);
-  l_bind         : (tscheme * tscheme);
-  l_subcomp      : (tscheme * tscheme);
-  l_if_then_else : (tscheme * tscheme);
-
+  l_bind         : (tscheme * tscheme * option indexed_effect_combinator_kind);
```

Comment:
Indexed effect combinators maintain their kinds (substitutive or ad_hoc) (kinds are inferred in `TcEffect` at the time of typechecking the indexed effect.